frontend/projects/upgrade/src/app/shared-standalone-component-lib/components/common-tag-input/common-tag-input.component.ts

```diff
@@ -42,42 +47,145 @@ import { TranslateModule } from '@ngx-translate/core';
       multi: true,
     },
   ],
-  imports: [CommonModule, MatChipsModule, MatFormFieldModule, MatIconModule, MatInputModule, TranslateModule],
+  imports: [
+    CommonModule,
+    MatChipsModule,
+    MatFormFieldModule,
+    MatIconModule,
+    MatInputModule,
+    TranslateModule,
+    CommonImportContainerComponent,
+  ],
 })
 export class CommonTagsInputComponent implements ControlValueAccessor, OnInit {
-  showExportIcon = false;
-  showImportIcon = false;
-  @Input() actionButtons = false;
-  @Output() actionButtonClicked = new EventEmitter<void>();
-
+  @Input() inputType: CommonTagInputType = CommonTagInputType.TAGS;
+  @Input() label = '';
+  @Input() placeholder = '';
+  @Input() forceValidation = false;
+  @Output() downloadRequested = new EventEmitter<string[]>();
+
+  tagsExist = false;
+  isTouched = false;
+  showImportHelper = false;
+  private importFailedSubject = new BehaviorSubject<boolean>(false);
+  importFailed$ = this.importFailedSubject.asObservable();
   isChipSelectable = false;
   isChipRemovable = true;
   addChipOnBlur = true;
   readonly separatorKeysCodes: number[] = [ENTER, COMMA];
 
   tags = new FormControl<string[]>([]);
+  CommonTagInputType = CommonTagInputType;
 
   ngOnInit(): void {
-    this.checkTagValue();
-  }
-
-  checkTagValue(): void {
     this.tags.valueChanges.subscribe((value) => {
-      if (this.actionButtons) {
-        // Update showExportIcon and showImportIcon based on tags value
-        if (value && value.length > 0) {
-          this.showExportIcon = true;
-          this.showImportIcon = false;
-        } else {
-          this.showImportIcon = true;
-          this.showExportIcon = false;
-        }
-      }
+      this.tagsExist = value && value.length > 0;
     });
   }
 
-  onActionButtonClick(): void {
-    this.actionButtonClicked.emit();
+  onFocus() {
+    this.onTouched();
+  }
+
+  onBlur() {
+    this.isTouched = true;
+    this.onTouched();
+  }
+
+  isInvalid(): boolean {
+    if (this.inputType !== CommonTagInputType.VALUES) {
+      return false;
+    }
+
+    return (this.isTouched || this.forceValidation) && (!this.tags.value || this.tags.value.length === 0);
+  }
+
+  onActionButtonClick(event: MouseEvent): void {
+    event.preventDefault();
+    event.stopPropagation();
+
+    if (this.tagsExist) {
+      this.downloadRequested.emit(this.tags.value || []);
+    } else {
+      this.importFailedSubject.next(false);
+      this.showImportHelper = true;
+    }
+  }
+
+  onCloseButtonClick(event: MouseEvent): void {
+    event.preventDefault();
+    this.showImportHelper = false;
+  }
+
+  handleFilesSelected(files: File[]) {
```

Comment:
I think the main thing hanging me up from approving is just how complicated it is to read this part of the code. It was helpful in the demo to understand what this code is actually doing, so I think it's just a readability thing. I just don't know if I've ever seen anyone in a project like this actually have to create an Observable from scratch, usually we don't _need_ to but maybe that's cool, as long as everyone can understand what's going with this.

I think what would help a lot be to just simply break out the anonymous callback functions into their own named functions. Aka like the `reduce` callback, the `next` callback, the `onload` functions can all be broken out and named and commented a little. The error callbacks could be too but those are pretty simple.